language/move-lang/src/interface_generator.rs

```diff
@@ -0,0 +1,284 @@
+    }
+    if !members.is_empty() {
+        members.push("".to_string());
+    }
+
+    let mut public_funs = module
+        .function_defs()
+        .iter()
+        .filter(|fdef| fdef.is_public)
+        .peekable();
+    if public_funs.peek().is_some() {
+        members.push(format!("    {}", DISCLAIMER));
+    }
+    for public_fdef in public_funs {
+        members.push(write_function_def(&mut context, public_fdef));
+    }
+    if !members.is_empty() {
+        members.push("".to_string());
+    }
+
+    let has_uses = !context.uses.is_empty();
+    for (module_id, alias) in context.uses {
+        if module_id.name().as_str() == alias {
+            push_line!(
+                out,
+                format!(
+                    "    use {}::{};",
+                    Address::new(module_id.address().to_u8()),
+                    module_id.name()
+                )
+            );
+        } else {
+            push_line!(
+                out,
+                format!(
+                    "    use {}::{} as {};",
+                    Address::new(module_id.address().to_u8()),
+                    module_id.name(),
+                    alias
+                )
+            );
+        }
+    }
+    if has_uses {
+        push_line!(out, "");
+    }
+
+    if !members.is_empty() {
+        push_line!(out, members.join("\n"));
+    }
+    push_line!(out, "}");
+    push_line!(out, "}");
+    Ok((id, out))
+}
+
+struct Context<'a> {
+    module: &'a CompiledModule,
+    uses: BTreeMap<ModuleId, String>,
+    counts: BTreeMap<String, usize>,
+}
+
+impl<'a> Context<'a> {
+    fn new(module: &'a CompiledModule) -> Self {
+        Self {
+            module,
+            uses: BTreeMap::new(),
+            counts: BTreeMap::new(),
+        }
+    }
+}
+
+const DISCLAIMER: &str = "// NOTE: Functions are 'native' for simplicity. They may or \
+                          may not be native in actuality.";
+
+fn write_struct_def(ctx: &mut Context, sdef: &StructDefinition) -> String {
+    let mut out = String::new();
+
+    let shandle = ctx.module.struct_handle_at(sdef.struct_handle);
+    let resource_mod = if shandle.is_nominal_resource {
+        "resource "
+    } else {
+        ""
+    };
+
+    push_line!(
+        out,
+        format!(
+            "    {}struct {}{} {{",
+            resource_mod,
+            ctx.module.identifier_at(shandle.name),
+            write_type_paramters(&shandle.type_parameters),
+        )
+    );
+
+    let fields = match &sdef.field_information {
+        StructFieldInformation::Native => {
+            push!(out, "    }");
+            return out;
+        }
+        StructFieldInformation::Declared(fields) => fields,
+    };
+    for field in fields {
+        push_line!(
+            out,
+            format!(
+                "        {}: {},",
+                ctx.module.identifier_at(field.name),
+                write_signature_token(ctx, &field.signature.0),
+            )
+        )
+    }
```

Comment:
Just some random thoughts, but I wonder if we would prefer translating struct defs to something more opaque (like the old native struct) if they ever become available in the future... or maybe not? Technically speaking fields are not accessible from outside module so there is no benefit for the Move program to know them, but maybe a smart compiler can leverage the extra knowledge about the layout in some way?

Reply:
I originally started with `native` structs (still supported in the source language compiler), but I figured seeing the fields could be helpful for trying to orient oneself when looking at on chain data

In other words, having the fields there might be helpful if you are using this in conjunction with something like the resource viewer

Reply:
Fair point. I was thinking of this more as a library producing outputs that get consumed by other tools, but now I can see its value to a human reader.